### PR TITLE
Add text highlight mode for text background

### LINKED ISSUE #468: Text - Highlight
## Feature Description

As an author I want to be able to highlight text

**Design and UX**
Textbox color s defaulted to none. As soon as you add a color, you should see option for textbox fill or highight. It is defaulted to fill. Heads up that I need to add additonal rules for how textbox color works with border radius and how to adjust line height for textbox highlight state but this shouldn't block initial work on just getting the basic functionality in.

image 

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria

1. The author is able to add highlights to text (not the entire box)

### PR DESCRIPTION
**Pull request by obetomuniz**
Resolves #468 

Kapture-2020-03-20-at-18 46 16

### CODE REVIEW

File: assets/src/edit-story/components/panels/textStyle/color.js
Comment (barklund) [inline assets/src/edit-story/components/panels/textStyle/color.js]: So it can be `'fill'`, `'highlight'` or `''`? Does that make sense? Wouldn't a boolean, e.g. `isBackgroundHighlight` be more appropriate defaulting to `false`?

File: assets/src/edit-story/components/panels/textStyle/color.js
Comment (barklund) [inline assets/src/edit-story/components/panels/textStyle/color.js]: What happens if multiple text fields with different settings are selected?

File: assets/src/edit-story/elements/text/display.js
Comment (barklund) [inline assets/src/edit-story/elements/text/display.js]: Surprisingly good support for this, but why isn't it auto-prefixed?

File: assets/src/edit-story/elements/text/display.js
Comment (barklund) [inline assets/src/edit-story/elements/text/display.js]: I'm not sure I understand this. Why do we need a clone?

File: assets/src/edit-story/components/panels/textStyle/color.js
Comment (barklund) [inline assets/src/edit-story/components/panels/textStyle/color.js]: I would expect to be able to toggle these with arrow keys too, when focused. Should be quite simple with `useKeyDownEffect`.

File: assets/src/edit-story/elements/text/output.js
Comment (barklund) [inline assets/src/edit-story/elements/text/output.js]: When saving, I get a warning in the console:

```
Warning: Unsupported vendor-prefixed style property webkitBoxDecorationBreak. Did you mean WebkitBoxDecorationBreak?
```

File: assets/src/edit-story/elements/text/display.js
Comment (barklund) [inline assets/src/edit-story/elements/text/display.js]: In the page preview, the text box doesn't render correctly (or at least not similarly to the on-canvas version:

Compare the initial text box in the page thumbnail in the carousel with the full background color to the highlight version after the switch (line-height seems to change completely breaking the layout?):

preview

File: assets/src/edit-story/elements/text/display.js
Comment (barklund) [reply assets/src/edit-story/elements/text/display.js]: Hm, it seems to be auto-prefixed and now actually duplicated?

Screen Shot 2020-03-23 at 6 23 04 PM

File: assets/src/edit-story/elements/text/output.js
Comment (barklund) [reply assets/src/edit-story/elements/text/output.js]: Not consistently though. Not sure why it came about.

**Review by barklund — COMMENTED** [review]
Looks good. I just need to understand the need for the clone in particular.

File: assets/src/edit-story/components/panels/textStyle/color.js
Comment (barklund) [reply assets/src/edit-story/components/panels/textStyle/color.js]: Just noticed [this demo](https://xwp.slack.com/archives/CCCAUAH6F/p1585060989330600?thread_ts=1585031022.322800&cid=CCCAUAH6F). We **will** need three states for background fill type, with "none" being default. Please add this extra button for transparent fill and maybe create some constants for the 3 values.

Also, we then need a data migration setting background fill type for all previous text elements based on whether or not they have a background color. If background is currently null or transparent color, set to `NONE`, otherwise set to `FILL`.

Comment (barklund) [reply assets/src/edit-story/components/panels/textStyle/color.js]: And of course use these new values in output component too.

File: assets/src/edit-story/elements/text/output.js
Comment (obetomuniz) [reply assets/src/edit-story/elements/text/output.js]: In this case, the `-webkit-box-decoration-break` is not being autogenerated. Anyway, just fixed the `w` lowercase 😄

File: assets/src/edit-story/elements/text/display.js
Comment (obetomuniz) [reply assets/src/edit-story/elements/text/display.js]: Actually, removing that and improving the example, I am being able to remove that clone node.

File: assets/src/edit-story/elements/text/display.js
Comment (obetomuniz) [reply assets/src/edit-story/elements/text/display.js]: I did not work yet with that mini live preview. Thanks for point this out anyway 👍

File: assets/src/edit-story/elements/text/display.js
Comment (obetomuniz) [reply assets/src/edit-story/elements/text/display.js]: I think that I was getting some caching issue and I did notice that, but yeah, this explicit declaration is creating duplicates for this case. Just removing. Thanks 👍

File: assets/src/edit-story/components/panels/textStyle/color.js
Comment (obetomuniz) [reply assets/src/edit-story/components/panels/textStyle/color.js]: Sure! I will take a look at examples using `useKeyDownEffect ` to achieve this. Just to clarify, using <kbd><</kbd> or <kbd>></kbd> you want to be able to switch between both, right?

File: assets/src/edit-story/components/panels/textStyle/color.js
Comment (obetomuniz) [reply assets/src/edit-story/components/panels/textStyle/color.js]: Yeah, sounds really reasonable add a NONE state. About data migration, what do you mean is to add a hardcoded condition based on the rules suggested above by you?

File: assets/src/edit-story/components/panels/textStyle/color.js
Comment (obetomuniz) [reply assets/src/edit-story/components/panels/textStyle/color.js]: Not sure yet, what do you suggest based on past elements? Maybe @samitron7 can bring some clarification on this behavior.

File: assets/src/edit-story/elements/text/output.js
Comment (obetomuniz) [reply assets/src/edit-story/elements/text/output.js]: 🤔 Looks like that our `npm run lint:js:fix` task before a commit is reverting always an uppercase `W` to `w`. I will take a look to disable that for this case.

File: assets/src/edit-story/components/panels/textStyle/color.js
Comment (barklund) [reply assets/src/edit-story/components/panels/textStyle/color.js]: Yes, add a new migration to `migrations` in the root folder and make sure to set proper background fill type on all old elements in old stories.

File: assets/src/edit-story/components/panels/textStyle/color.js
Comment (barklund) [reply assets/src/edit-story/components/panels/textStyle/color.js]: Elsewhere our behavior is to show "(multiple)", but how to do that for a toggle like this. It's kind of like selecting two text fields, one of them bold, the other not.

I guess the common behavior in such a case is to mark neither button as selected (as the state is MULTIPLE_STATE and not any of the valid options), and when clicking any background fill button, make all the textfields sync'ed.

File: assets/src/edit-story/components/panels/textStyle/color.js
Comment (barklund) [reply assets/src/edit-story/components/panels/textStyle/color.js]: I wouldn't expect it to cycle. Just pressing <kbd>left</kbd> moves to the next button to the left (and selects it). Clicking <kbd>left</kbd> when already at the left-most button does nothing.

File: assets/src/edit-story/components/panels/textStyle/color.js
Comment (dvoytenko) [inline assets/src/edit-story/components/panels/textStyle/color.js]: Consistently with the `BACKGROUND_TEXT_MODE`, let's call this property `backgroundTextMode`. Also please add it to the `types.js` for text elements.

File: assets/src/edit-story/components/panels/textStyle/color.js
Comment (dvoytenko) [inline assets/src/edit-story/components/panels/textStyle/color.js]: Does the empty string actually have meaning for us? Or is it always `BACKGROUND_TEXT_MODE.NONE`? If so, let's use the `BACKGROUND_TEXT_MODE.NONE` value? Alternatively it might be cleaner to do update on the `true` value only. E.g.:

```
onChange({value} => value && pushUpdate({backgroundMediaMode: BACKGROUND_TEXT_MODE.HIGHLIGHT}, true))
```

File: assets/src/edit-story/elements/text/display.js
Comment (dvoytenko) [inline assets/src/edit-story/elements/text/display.js]: It's really underseriable to change refs. Any way this can be done differently? Some top-level container? Or we can apply the exact same DOM structure of `p > span`, but customize the CSS for both `p and span` when a specific background mode is used.

File: assets/src/edit-story/elements/text/output.js
Comment (dvoytenko) [inline assets/src/edit-story/elements/text/output.js]: I believe this should still use `className="fill"`.

File: assets/src/edit-story/elements/text/output.js
Comment (dvoytenko) [inline assets/src/edit-story/elements/text/output.js]: Is this correct? The `bgColor` if I read this right also sets `background-image` and might be some more. Perhaps we shouldn't set `...bgColor` at all for this mode and fork these two styles a bit higher?

File: assets/src/edit-story/elements/text/index.js
Comment (dvoytenko) [inline assets/src/edit-story/elements/text/index.js]: I think default should be FILL

**Review by dvoytenko — CHANGES_REQUESTED** [review]

**Comment by pbakaus** [comment]
hi Beto, excited to see this taking shape! A couple of issues that I'd love to get addressed:

1. When "Highlight" fill style is active, the padding input should affect the inner padding of the span, *not* the bounding box. While this is technically a different padding, it should feel the most intuitive.
2. I think we should try to calculate the line-height in this mode based on how I've done it in the [codepen](https://codepen.io/pbakaus/pen/rNVddXg) (see highlight spacing). A line height of "1", in highlight fill mode, shall mean "no spacing between lines". This is a little tricky to figure out, so can be a follow up PR.
3. In my codepen, I purposely clone the element and stack it on top of each other at the exact same position (it's a little too hacky in my codepen, obviously for your purposes use some sort of wrapper). This is important, as with overlapping lines (a desired effect, see codepen and set spacing to -0.3em), the letters otherwise get cut off.
4. Because 3) will be a common effect people use, we should not allow them to lower the opacity when the lines overlap - so please include some logic that disables opacity temporarily (disables the opacity field, but keeps the value) in that case.

Feel free to ping me if you have any questions.

File: assets/src/edit-story/elements/text/display.js
Comment (obetomuniz) [reply assets/src/edit-story/elements/text/display.js]: Interesting. I will try this `top-level container` approach for this case. Thanks, @dvoytenko !

**Comment by barklund** [comment]
@googlebot I consent.

**Comment by barklund** [comment]
Alright, it's getting a lot closer, but still a bit hacky. `-webkit-box-decoration-break` isn't rendered perfectly in Chrome and requires some CSS hacks to properly box in (it doesn't subtract padding properly, thus a need for a third element).

Here's a semi-transparent conic gradient highlight fill with negative line height and lots of padding displayed in the editor (_left_) and in the rendered amp story (_right_). Notice how the line breaks aren't calculated similarly between the two versions resulting in layout differences - but other than that it looks pretty good:

Screen Shot 2020-03-25 at 10 02 08 PM

@pbakaus feel free to play around with the most recent version of this PR to get a feel for it. I tried to replicate your desires about both line height and padding.

* Line height for highlight text fields is now relative - a line height of 0 is normal distance, negative overlaps the lines and positive adds space between the lines. It's not perfect, but pretty close.
* Padding is not added to the text box itself, only to the box decorated lines inside the inner span.

File: assets/src/edit-story/elements/text/output.js
Comment (barklund) [reply assets/src/edit-story/elements/text/output.js]: Fixed.

File: assets/src/edit-story/elements/text/display.js
Comment (barklund) [reply assets/src/edit-story/elements/text/display.js]: I did not address this one, though. Is it necessary?

**Comment by pbakaus** [comment]
@barklund thanks! couple things:

1.  normal distance of line-height should be 1, not 0. The jump between fill and highlight is currently pretty jarring because of that. Bonus points if you can calculate between them so that the text stays where it is..
2. with line-height zero, the box is clipped, as the span now moves into negative y territory (look at the rounded corners)
3. When increasing the line-height, a visible gap at the top and bottom around the blue outline appears (the box being bigger than the span). This isn't desired.
4. Padding works pretty nice! But does increase the gap in #3 even more.

Getting closer! This is a fairly critical feature for the alpha, so I'm considering making an exception to get this in, but ultimately will leave this decision to @swissspidy.

**Comment by pbakaus** [comment]
Forgot an important one:

5. Text box resizing is super out of whack when highlight is active. Should be fixed.

**Comment by barklund** [comment]
> Getting closer! This is a fairly critical feature for the alpha, so I'm considering making an exception to get this in, but ultimately will leave this decision to @swissspidy.

I'll give it one more shot.

> 1. normal distance of line-height should be 1, not 0. The jump between fill and highlight is currently pretty jarring because of that. Bonus points if you can calculate between them so that the text stays where it is..

Given that padding acts very differently in the two modes (per box vs per line), this (text stays the same when switching mode) won't be possible for any multi-line text field with non-zero padding.

> 5. Text box resizing is super out of whack when highlight is active. Should be fixed.

Yes, that's a tough one to do anything about in the current setup, I think - at least tonight. I'd rather leave that for a new ticket to be fixed later.

**Comment by pbakaus** [comment]
> > Getting closer! This is a fairly critical feature for the alpha, so I'm considering making an exception to get this in, but ultimately will leave this decision to @swissspidy.
> 
> I'll give it one more shot.
> 
> > 1. normal distance of line-height should be 1, not 0. The jump between fill and highlight is currently pretty jarring because of that. Bonus points if you can calculate between them so that the text stays where it is..
> 
> Given that padding acts very differently in the two modes (per box vs per line), this (text stays the same when switching mode) won't be possible for any multi-line text field with non-zero padding.

Yeah fair point. I think it will be a lot less jarring with the change from from 0 to 1. If it's still very jarring then, we can consider making it "dumb" again (no change to line-height, no recalc etc.).

> 
> > 1. Text box resizing is super out of whack when highlight is active. Should be fixed.
> 
> Yes, that's a tough one to do anything about in the current setup, I think - at least tonight. I'd rather leave that for a new ticket to be fixed later.

SGTM.

**Comment by barklund** [comment]
> 2. with line-height zero, the box is clipped, as the span now moves into negative y territory (look at the rounded corners)

That's actually a consequence of keeping the lines matching in size and placement between the modes. With positive line height, there will be space above the first line box, with negative the first line box will be cut short from the top. BUT, the first line will correctly match the position of the first line in regular fill mode. If we start move the first line up or down to make the box flux with the outer box top, then the first line won't match at all when switching modes.

I've just pushed a new version aligning things better.

If you have 0 padding, the lines will match perfectly now when switching mode (ignore the box changing size for now). If you have non-zero padding, the first line will match perfectly when switching mode, but subsequent lines will be pushed down ever more due to padding.

How do we reconcile these two visual wants:

* Lines should match when switching from fill to highlight.
* When in highlight mode, the first line box should be align perfect top with the outer box.

Also, how do I actually calculate this offset? It's some calculation in line height and padding, but I can't really seem to grasp it. I can get close, but it'll still either gap or cut it short.

**Comment by pbakaus** [comment]
@barklund this is really nice now! No concerns from my side anymore re: functionality at this moment. Please resolve conflicts and then lets get this baby merged if @dvoytenko approves.

We can then complete it with two follow-up issues / PRs:

1. Smooth highlight resizing
2. Perfect highlight formula to produce exact bounding box

@jauyong can you help file these new issues and assign to Beto and Morten?

**Comment by pbakaus** [comment]
@barklund found a bug with the outer part of the bounding box when padding is used:

image

File: assets/src/edit-story/elements/text/display.js
Comment (pbakaus) [reply assets/src/edit-story/elements/text/display.js]: @dvoytenko @barklund what's the resolution here? Dima, can you advise on whether this should still be addressed before merging?

**Comment by barklund** [comment]
I been struggling with this one all afternoon. I simply cannot get the word breaks to align in different calculations. It just can't happen. A scaled text box does not break at the same breakpoints - fonts just don't work like that.

In this video here:
https://youtu.be/l1ZP2ttOxzE

the left-most box is a visualization of the output render at 30% scale (it's calculated at the full 1080x1920 px size, then scaled down using `transform:scale(.3)`). The middle render use scaled pixel values (that scale to the current canvas size, which is some fraction of the full 1080x1920). Even though the pixels aren't rounded, there's still a difference. Notice how some word break happen in the right place, and some happen slightly earlier on the left version (the transformed one). I even found some instances where the word breaks happened first on the left for some lines and first on the middle for other lines.

The right-most version is the rendered story using percentages of the whole story width - directly calculated from the 1080x1920px valued just by ratios instead.

As you can see, there will be cases where the calculated height doesn't match the visual height. I can make sure it errs on either side though, so it's always guaranteed that all the lines will fit inside the visual box (but the box might be a line or two too long), or that the box will never be too long (but might cut off the last line or two).

And this isn't only an issue for highlight text of course. I can find the same pixel-tight deviations for regular text as well.

The only "real" solution as I see it is to manually find line breaks in one of these instances and calculate number of lines from that - and then use explicit line breaks in the real output to match this calculation. We can calculate line breaks by pasting word for word into a fixed-width box and see when the height shifts or we can leverage `CanvasRenderingContext2D.measureText()`.

@pbakaus, what's your call here? Do you have prior experience? Should we manually split the lines to ensure similarity or can we live with the slight shifts? The cut-off text is probably a big no-go here, right?

**Comment by pbakaus** [comment]
@barklund my condolences, this is indeed a tricky thing to do, and I've faced similar issues before. The browser's layout process works in mysterious ways sometimes.. So a couple things:

1. cut-off text is definitely a big no-go.
2. I think you are on the right track. The minuscule differences are usually related to scaling and rounding, and so slightly correcting the output in either direction should do the job.

Here's what I'd do: After resizing etc. (when you're ready to 'commit' a value), measure the total height of the span (fine to use offsetHeight or something here, causes layout but that's the lesser evil in this scenario) to find out how many lines you got, compare it to the total height of the percentage-based/resized 'finalized state' off screen, then, if the line count if off by one in either direction, add a 'padding' to the calculated float (experiment here, ~0.1 or ~0.01% might be enough) to correct.

If it helps, I'm happy to jump on a call for a pair programming / troubleshooting session or whatever. Good luck!

**Comment by barklund** [comment]
@pbakaus, the trick is though, that I can't test render the final story. At least we don't do that right now. I can only story some variables, that will be used to render the final story, but I can't actually "test run" it and see how long the text will be, how many lines will fit.

Also, this won't ever be static anyway, as the rendered story has width, height and font size in percent of the overall page size. Though these will relatively stay the same as the page resizes, the line breaks **will** change in the rendered story when the page resizes - that's unavoidable. I don't think we can find some magic relative width and height, that'll make all texts break correctly for all page sizes.

As far as I can see, we can only do two things to "fix" this:

1. Insert manual line breaks to make sure the lines always break exactly where we want them to break.
2. Remember the number of lines that's supposed to be in the output, send that to the rendered story and as an attribute of the textfield and in the rendered story add some AMP-compatible script, that'll for initial render and any resize validate, that the text field does in deed contain the required number of lines (and if not decrement/increment the width slightly until it does).

**Comment by pbakaus** [comment]
@barklund I'd be very surprised if 1) doesn't produce an accessibility / screen reader nightmare.. so I'm definitely leaning towards 2. But I'm not convinced that we can't simulate the final output under the hood in the editor. What stops us from recreating the conditions in the final output?

**Comment by barklund** [comment]
> I'd be very surprised if 1) doesn't produce an accessibility / screen reader nightmare.

I actually just want to insert `\n` and do `white-space: pre`. That shouldn't impact accessibility at all, should it?

> What stops us from recreating the conditions in the final output?

The output is not just one size. The output can be rendered in any size, and we'd have to check that the line breaks are correct for all of them - or at least a representative number of them. That sounds expensive!

**Comment by barklund** [comment]
> > What stops us from recreating the conditions in the final output?
> 
> The output is not just one size. The output can be rendered in any size, and we'd have to check that the line breaks are correct for all of them - or at least a representative number of them. That sounds expensive!

@pbakaus Hm, it seems the rendered story will always break lines at the same points at various viewport sizes - or at least I can't seem to find any combination of text size, font and width that'll make any line break differently at different viewport sizes. So that's a huge first step at least.

But in order to test how a rendered story will render a given text field, we need to create an iframe with an AMP story inside and measure it here? Or can we create "normal" html elements and assume they'd render identically?

I'm not completely aware how much "magic" (if any) that AMP adds to the browser rendering engine.

**Comment by pbakaus** [comment]
@barklund I don't think there's a crazy amount of magic, it depends more on the way resizing is done (transforms vs. pecentage vs. font-size vs. zoom). At this point, I'd say edge cases are OK, I'm mostly looking for a good enough approximation. Let's merge this soon and then polish later?

@dvoytenko I'd say ok to go ahead and review again.

File: assets/src/edit-story/elements/text/display.js
Comment (dvoytenko) [inline assets/src/edit-story/elements/text/display.js]: Is `padding` always defined? I ran into some cases and have been using `padding?.horizontal` elsewhere. But I might have just run into some old cases.

File: assets/src/edit-story/elements/text/display.js
Comment (dvoytenko) [inline assets/src/edit-story/elements/text/display.js]: It's weird that the `ref` is set on both `HighlightElement` instances.

File: assets/src/edit-story/elements/text/display.js
Comment (dvoytenko) [inline assets/src/edit-story/elements/text/display.js]: Does it have to be `display: inline-block`?

File: assets/src/edit-story/elements/text/frame.js
Comment (dvoytenko) [inline assets/src/edit-story/elements/text/frame.js:44]: Are there any special highlight work in frame? We use an invisible text to figure out which part of an element is selected for editor.

File: assets/src/edit-story/elements/text/output.js
Comment (dvoytenko) [inline assets/src/edit-story/elements/text/output.js]: Hmm. Let's at least document the intent here?

File: assets/src/edit-story/elements/text/output.js
Comment (dvoytenko) [inline assets/src/edit-story/elements/text/output.js]: One of these should be `aria-hidden`, right?

File: assets/src/edit-story/elements/text/output.js
Comment (dvoytenko) [inline assets/src/edit-story/elements/text/output.js]: Nit, but took me a second to figure out that this is `(padding.vertical/width) * 100`. Otherwise, I was trying to figure out where `padding.vertical` is recalculated to the base of 0-1.

File: assets/src/edit-story/migration/migrations/v0010_setBackgroundTextMode.js
Comment (dvoytenko) [inline assets/src/edit-story/migration/migrations/v0010_setBackgroundTextMode.js]: Does this apply to non-text as well?

**Review by dvoytenko — CHANGES_REQUESTED** [review]

File: assets/src/edit-story/elements/text/index.js
Comment (dvoytenko) [inline assets/src/edit-story/elements/text/index.js]: Also, it sounds like `NONE` should be default?

**Review by dvoytenko — CHANGES_REQUESTED** [review]

File: assets/src/edit-story/elements/text/display.js
Comment (barklund) [reply assets/src/edit-story/elements/text/display.js]: Yes, it's necessary here. It can't be either `inline` or `block`, as it has to mediate the white space around the element correctly - there's some browser quirks around padding calculation for `box-decoration-break`.

File: assets/src/edit-story/elements/text/frame.js
Comment (barklund) [reply assets/src/edit-story/elements/text/frame.js:44]: Not completely sure what you mean here. Currently there's no special handling in the frame for a highlight text element. What should happen in the frame, if things were done correctly?

File: assets/src/edit-story/elements/text/display.js
Comment (barklund) [reply assets/src/edit-story/elements/text/display.js]: Fixed in 25ff1b28.

File: assets/src/edit-story/elements/text/display.js
Comment (barklund) [reply assets/src/edit-story/elements/text/display.js]: Very weird! Fixed in 62e55403.

File: assets/src/edit-story/elements/text/output.js
Comment (barklund) [reply assets/src/edit-story/elements/text/output.js]: Redid this completely in 3dc9ab41. It's [a simpler calculation now](https://github.com/google/web-stories-wp/pull/684/files#diff-f364e182c74927d08611569486222239R69-R76), that makes more sense.

File: assets/src/edit-story/elements/text/output.js
Comment (barklund) [reply assets/src/edit-story/elements/text/output.js]: Yes, of course!

File: assets/src/edit-story/elements/text/index.js
Comment (barklund) [reply assets/src/edit-story/elements/text/index.js]: Fixed in f1e5c558.

**Comment by barklund** [comment]
I feel this is working pretty well now. There are many cases where line breaks occur differently in the editor preview vs the rendered story, but that's very hard to do anything about IMO. I've made sure that the text field is always at least as big as the contents, but sometimes it's too big (the test render has an extra line, thus room for an extra line in the preview), but that's erring on the right side, I feel.

I'll fix the edit component for highlight now. I don't really understand the purpose of the frame component, so not sure what needs fixing for this one.

@dvoytenko, PTAL again.

**Comment by barklund** [comment]
I simply cannot get highlight stying to work inside the editor component (draft.js). I don't really know what to do here. I have now made sure the edit component uses more or less the same line height as the display one, but there's still a lot of visual differences between editing and the resulting look. It's not a good user experience. but I recommend we merge this as-is and follow up asap.

Maybe someone else has better ideas for how to solve these last remaining issues?

File: assets/src/edit-story/elements/text/frame.js
Comment (dvoytenko) [reply assets/src/edit-story/elements/text/frame.js:44]: It'd be ok to move it to a separate pull request, as long as we don't forget about this entirely. But this is actually originally your code. We generally try to make sure the text in the frame is displayed (but not visible) the same way as display component to ensure that selection logic is correct. See this: https://github.com/google/web-stories-wp/blob/c6af16cfc77e8fc02efe215358a20e8dc8fd3f64/assets/src/edit-story/elements/text/frame.js#L111-L114

**Review by dvoytenko — APPROVED** [review]
I'm ok with this merging. IMHO let's look into edit mode and frame in separate pull requests. Please file the bugs though. Also, v0010 migration should now be bumped +1.

**Review by wassgha — APPROVED** [review]

**Pull request merged**
